platform-sdk/docs/proposals/state-lifecycle/proposal.md

```diff
@@ -33,66 +33,76 @@ These methods correspond to the state lifecycle as follows:
-- Loading snapshots from the disk.
+Therefore, `swirlds-state-api` needs a `StateLifecycleManager` interface with an implementation `StateLifecycleManagerImpl` in `swirlds-state-impl`  that will have the following responsibilities:
+- Take initial state as a starting point. It can be a genesis state or a state loaded from a snapshot.
+- Keep track of the references to the latest immutable state and the latest mutable state.
+- Restrict mutability to a single state. That is, only one state object should be mutable at any given time.
+
+The interface will look something like this:
+
+```java
+public interface StateLifecycleManager {
+    void setInitialState(State initialState);
+
+    State getLatestImmutableState();
 
-Additionally, there should be a configuration parameter that limits the number of states kept in memory. When the limit is reached, the oldest state should be evicted.
-Currently, we rely on the implicit garbage collection. We have a reference count mechanism to achieve this.
-We want to get rid of this mechanism and rely on the explicit eviction mechanism.
+    State getMutableState();
+}
+```
+
+**Imporatant**: It's up to the client code to make sure that the state is released once it's no longer needed.
+Without this, the state will not be evicted from memory and eventually the application will run out of memory.
 
 The following diagram illustrates an example of the states in memory throughout their lifecycles:
 
-[![Multistate management](multi-state-management.png)](multi-state-management.png)
+[![Multistate management](multi-states.svg)](multi-states.svg)
+
+Also, `swirlds-state-api` needs a `SnapshotManager` interface with an implementation `SnapshotManagerImpl` in `swirlds-state-impl` that will have the following responsibilities:
+- Load snapshots from the disk. A state loaded from a snapshot must be immutable and available for look up by the round number.
+- Evict the loaded state by the round number when the state is no longer needed.
+
+This functionality is separated from the `StateLifecycleManager` for the following reasons:
+- a state loaded from a snapshot is _always_ immutable, and therefore there is no notion of mutability here.
+- the mechanism of maintaining the loaded states in memory is different. This functionality doesn't rely on the reference counting.
+Instead, it has explicit methods to load a state from a snapshot, and evict the state by the round number when it's no longer needed.
 
 The interface will look something like this:
 
 ```java
-public interface StateManager {
-    State createEmptyState();
-    
-    State getLatestImmutableState();
-    
-    State getMutableState();
-    
-    State getState(long round);
-    
-    void evictState(long round);
-    
+public interface SnapshotManager {
+
+    State getLoadedState(long round);
+
     State loadSnapshot(Path sourcePath);
+
+    void evictLoadedState(long round);
 }
```

Comment:
I don't agree with the splitting of methods this way. The idea of the snapshot manager was to deal with instances of `Snapshot` which will be a new class that loads minimal information about a state from disk like the hashes that can be used to produce a merkle proof, but not load the entire state into memory. Loading an entire state into memory and providing it as a `State` object is a function of a `StateManager` or `States` object.

The `loadSnapshot` method should probably be `loadStateFromDisk` to differentiate loading the minimal data from loading the full state into memory.

Also, I don't see a reason for the `evict` or `getLoadedState` method anymore. When a state's references go to zero, it will should be GCed. And if the user of the module need to hang onto a state for collecting signatures or anything like that, it's going to have to keep track of the reference itself to prevent it from being GCed, therefore this method is not needed.

Reply:
Right now we don't maintain a list of the loaded states. There is no such functionality - we don't have a class that has `State getLoadedState(long round)` method. If we intend to have such a method, there should be some way to get rid of the states that are no longer needed, manual or automatic.